var/spack/repos/builtin/packages/openblas/package.py

```diff
@@ -98,8 +98,6 @@ class Openblas(MakefilePackage):
     # https://github.com/xianyi/OpenBLAS/pull/1703
     patch('openblas-0.3.2-cmake.patch', when='@0.3.1:0.3.2')
 
-    parallel = False
```

Comment:
There's something strange going on: OpenBLAS' GNU make based system has internal parallelism and actually breaks with `-jN` (N > 1) but it obviously doesn't work when building with Spack. Will investigate further and remove this patch from this PR for now.

Furthermore, when comparing to the way Gentoo builds it (https://gitweb.gentoo.org/proj/sci.git/tree/sci-libs/openblas/openblas-0.2.20.ebuild), we might also have to set `NUM_THREADS` to avoid limitting the max num threads for OpenMP to the build machine.

Reply:
ok, it seems that OpenBLAS 0.3 at least should work with parallel builds, see the following commit on openBLAS: https://github.com/xianyi/OpenBLAS/pull/829
For 0.3.x I therefore removed the Spack patch `make.patch` and use the official openBLAS flag to disable openBLAS' internal parallel build and enable parallel build.

Reply:
Thanks, can you check cp2k as well?

Reply:
working in the group developing CP2K I can assure you that we always supported parallel builds ;-) Look for the `-j ...` parameter in the logs (`Status` column) on our dashboard at https://dashboard.cp2k.org/

Reply:
Huh. It looks like this has been present since @alalazo's original package in #1383. @alalazo any comments?

Reply:
I did a few tests and it looks like the "parallel = False" is required for our our systems with arch =
linux-rhel6-x86_64. The build fails for version 0.3.3 and for version 0.2.19 with the link errors below. With arch = linux-rhel7-x86_64 and arch = linux-rhel7-ppc64le, I can build OK without "parallel = False". I don't know if the architectural difference is tied to the version of make or something else. I would be happy to see a solution that allows openblas to build in parallel when it won't break the spack build!


==> Error: ProcessError: Command exited with status 2:                          
    'make' '-j16' 'CC=/nfs/tmp2/lee218/delete/obrien_spack/lib/spack/env/gcc/gcc' 'FC=/nfs/tmp2/lee218/delete/obrien_spack/lib/spack/env/gcc/gfortran' 'MAKE_NO_J=1' 'NO_AVX2=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'libs' 'netlib' 'shared'         

77 errors found in build log:
     8012    perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 > linktest
             .c                                                                
     8013    /nfs/tmp2/lee218/delete/obrien_spack/lib/spack/env/gcc/gcc -O2 -DM
             AX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_AVX2
              -DNO_WARMUP -DMAX_CPU_NUMBER=16 -DASMNAME= -DASMFNAME=_ -DNAME=_ 
             -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -
             shared -o ../libopenblas_sandybridge-r0.2.19.so \                 
     8014       -Wl,--whole-archive ../libopenblas_sandybridge-r0.2.19.a -Wl,--no                                                                               
             -whole-archive \                                                   
     8015       -Wl,-soname,libopenblas.so.0 -lm -lgfortran -lm -lgfortran      
     8016    /nfs/tmp2/lee218/delete/obrien_spack/lib/spack/env/gcc/gcc -O2 -DM 
             AX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_AVX2 
              -DNO_WARMUP -DMAX_CPU_NUMBER=16 -DASMNAME= -DASMFNAME=_ -DNAME=_  
             -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  - 
             w -o linktest linktest.c ../libopenblas_sandybridge-r0.2.19.so -L/ 
             usr/lib/gcc/x86_64-redhat-linux6E/4.6.1 -L/usr/lib/gcc/x86_64-redh 
             at-linux6E/4.6.1/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../l 
             ib64 -L/usr/lib/gcc/x86_64-redhat-linux6E/4.6.1/../../..  -lgfortr 
             an -lm -lquadmath -lm -lc   && echo OK.                            
     8017    /tmp/ccTDOU7l.o: In function `main':                               
  >> 8018    linktest.c:(.text.startup+0x2f16): undefined reference to `clagge_ 
             '                                                                  
  >> 8019    linktest.c:(.text.startup+0x2f1d): undefined reference to `claghe_ 
             '                                                                  
  >> 8020    linktest.c:(.text.startup+0x2f24): undefined reference to `clagsy_ 
             '                                                                  
  >> 8021    linktest.c:(.text.startup+0x2f2b): undefined reference to `clahilb 
             _'                                                                 
  >> 8022    linktest.c:(.text.startup+0x2f32): undefined reference to `clakf2_ 
             '                                                                  
  >> 8023    linktest.c:(.text.startup+0x2f39): undefined reference to `clarge_ 
             '                                                                  
  >> 8024    linktest.c:(.text.startup+0x2f40): undefined reference to `clarnd_ 
             '                                                                  
  >> 8025    linktest.c:(.text.startup+0x2f47): undefined reference to `claror_ 
             '                                                                  
  >> 8026    linktest.c:(.text.startup+0x2f4e): undefined reference to `clarot_ 
             '                                                                  
  >> 8027    linktest.c:(.text.startup+0x2f55): undefined reference to `clatm1_ 
             '                                                                  
  >> 8028    linktest.c:(.text.startup+0x2f5c): undefined reference to `clatm2_ 
             '                                                                  
  >> 8029    linktest.c:(.text.startup+0x2f63): undefined reference to `clatm3_ 
             '                                                                  
  >> 8030    linktest.c:(.text.startup+0x2f6a): undefined reference to `clatm5_ 
             '                                                                  
  >> 8031    linktest.c:(.text.startup+0x2f71): undefined reference to `clatm6_ 
             '                                                                  
  >> 8032    linktest.c:(.text.startup+0x2f78): undefined reference to `clatme_ 
             '                                                                  
  >> 8033    linktest.c:(.text.startup+0x2f7f): undefined reference to `clatmr_ 
             '                                                                  
  >> 8034    linktest.c:(.text.startup+0x2f86): undefined reference to `clatms_ 
             '                                                                  
  >> 8035    linktest.c:(.text.startup+0x2f8d): undefined reference to `clatmt_ 
             '                                                                  
  >> 8036    linktest.c:(.text.startup+0x2ff6): undefined reference to `dlagge_ 
             '                                                                  
  >> 8037    linktest.c:(.text.startup+0x2ffd): undefined reference to `dlagsy_ 
             '                                                                  
  >> 8038    linktest.c:(.text.startup+0x3004): undefined reference to `dlahilb 
             _'                                                                 
  >> 8039    linktest.c:(.text.startup+0x300b): undefined reference to `dlakf2_ 
             '                                                                  
  >> 8040    linktest.c:(.text.startup+0x3012): undefined reference to `dlaran_ 
             '                                                                  
  >> 8041    linktest.c:(.text.startup+0x3019): undefined reference to `dlarge_ 
             '                                                                  
  >> 8042    linktest.c:(.text.startup+0x3020): undefined reference to `dlarnd_ 
             '                                                                  
  >> 8043    linktest.c:(.text.startup+0x3027): undefined reference to `dlaror_ 
             '                                                                  
  >> 8044    linktest.c:(.text.startup+0x302e): undefined reference to `dlarot_ 
             '                                                                  
  >> 8045    linktest.c:(.text.startup+0x3035): undefined reference to `dlatm1_ 
             '                                                                  
  >> 8046    linktest.c:(.text.startup+0x303c): undefined reference to `dlatm2_ 
             '                                                                  
  >> 8047    linktest.c:(.text.startup+0x3043): undefined reference to `dlatm3_ 
             '                                                                  
  >> 8048    linktest.c:(.text.startup+0x304a): undefined reference to `dlatm5_ 
             '                                                                  
  >> 8049    linktest.c:(.text.startup+0x3051): undefined reference to `dlatm6_ 
             '                                                                  
  >> 8050    linktest.c:(.text.startup+0x3058): undefined reference to `dlatm7_ 
             '                                                                  
  >> 8051    linktest.c:(.text.startup+0x305f): undefined reference to `dlatme_ 
             '                                                                  
  >> 8052    linktest.c:(.text.startup+0x3066): undefined reference to `dlatmr_ 
             '                                                                  
  >> 8053    linktest.c:(.text.startup+0x306d): undefined reference to `dlatms_ 
             '                                                                  
  >> 8054    linktest.c:(.text.startup+0x3074): undefined reference to `dlatmt_ 
             '                                                                  
  >> 8055    linktest.c:(.text.startup+0x30d6): undefined reference to `slagge_ 
             '                                                                  
  >> 8056    linktest.c:(.text.startup+0x30dd): undefined reference to `slagsy_ 
             '                                                                  
  >> 8057    linktest.c:(.text.startup+0x30e4): undefined reference to `slahilb 
             _'                                                                 
  >> 8058    linktest.c:(.text.startup+0x30eb): undefined reference to `slakf2_ 
             '                                                                  
  >> 8059    linktest.c:(.text.startup+0x30f2): undefined reference to `slaran_ 
             '                                                                  
  >> 8060    linktest.c:(.text.startup+0x30f9): undefined reference to `slarge_ 
             '                                                                  
  >> 8061    linktest.c:(.text.startup+0x3100): undefined reference to `slarnd_ 
             '                                                                  
  >> 8062    linktest.c:(.text.startup+0x3107): undefined reference to `slaror_ 
             '                                                                  
  >> 8063    linktest.c:(.text.startup+0x310e): undefined reference to `slarot_ 
             '                                                                  
  >> 8064    linktest.c:(.text.startup+0x3115): undefined reference to `slatm1_ 
             '                                                                  
  >> 8065    linktest.c:(.text.startup+0x311c): undefined reference to `slatm2_ 
             '                                                                  
  >> 8066    linktest.c:(.text.startup+0x3123): undefined reference to `slatm3_ 
             '                                                                  
  >> 8067    linktest.c:(.text.startup+0x312a): undefined reference to `slatm5_ 
             '                                                                  
  >> 8068    linktest.c:(.text.startup+0x3131): undefined reference to `slatm6_
             '
  >> 8069    linktest.c:(.text.startup+0x3138): undefined reference to `slatm7_
             '
  >> 8070    linktest.c:(.text.startup+0x313f): undefined reference to `slatme_
             '
  >> 8071    linktest.c:(.text.startup+0x3146): undefined reference to `slatmr_
             '
  >> 8072    linktest.c:(.text.startup+0x314d): undefined reference to `slatms_
             '
  >> 8073    linktest.c:(.text.startup+0x3154): undefined reference to `slatmt_
             '
  >> 8074    linktest.c:(.text.startup+0x31b6): undefined reference to `zlagge_
             '
  >> 8075    linktest.c:(.text.startup+0x31bd): undefined reference to `zlaghe_
             '
  >> 8076    linktest.c:(.text.startup+0x31c4): undefined reference to `zlagsy_
             '
  >> 8077    linktest.c:(.text.startup+0x31cb): undefined reference to `zlahilb
             _'
  >> 8078    linktest.c:(.text.startup+0x31d2): undefined reference to `zlakf2_
             '
  >> 8079    linktest.c:(.text.startup+0x31d9): undefined reference to `zlarge_
             '
  >> 8080    linktest.c:(.text.startup+0x31e0): undefined reference to `zlarnd_
             '
  >> 8081    linktest.c:(.text.startup+0x31e7): undefined reference to `zlaror_
             '
  >> 8082    linktest.c:(.text.startup+0x31ee): undefined reference to `zlarot_
             '
  >> 8083    linktest.c:(.text.startup+0x31f5): undefined reference to `zlatm1_
             '
  >> 8084    linktest.c:(.text.startup+0x31fc): undefined reference to `zlatm2_
             '
  >> 8085    linktest.c:(.text.startup+0x3203): undefined reference to `zlatm3_
             '
  >> 8086    linktest.c:(.text.startup+0x320a): undefined reference to `zlatm5_
             '
  >> 8087    linktest.c:(.text.startup+0x3211): undefined reference to `zlatm6_
             '
  >> 8088    linktest.c:(.text.startup+0x3218): undefined reference to `zlatme_
             '
  >> 8089    linktest.c:(.text.startup+0x321f): undefined reference to `zlatmr_
             '
  >> 8090    linktest.c:(.text.startup+0x3226): undefined reference to `zlatms_
             '
  >> 8091    linktest.c:(.text.startup+0x322d): undefined reference to `zlatmt_
             '
  >> 8092    collect2: ld returned 1 exit status
  >> 8093    make[1]: *** [../libopenblas_sandybridge-r0.2.19.so] Error 1
     8094    make[1]: Leaving directory `/tmp/lee218/spack-stage/spack-stage-Cn
             7NjO/OpenBLAS-0.2.19/exports'
  >> 8095    make: *** [shared] Error 2

Reply:
@lee218llnl did you do that test with the `package.py` from my PR? Because it should not change anything for OpenBLAS < 0.3.0, see for example https://github.com/spack/spack/pull/9308/commits/292acb0f07638228e62eaa718ff881cbb278d2fb#diff-63e8de85c24cd4293bbb252347d11cf3R104
For OpenBLAS >= 0.3 I removed the `make.patch` since OpenBLAS officially supports specifying make flags externally. If you can still reproduce with the latest version from my PR for OpenBLAS 0.3.3, can you please verify whether it fails with the exact same error message and otherwise post the error here as well?

Reply:
I still see the errors intermittently on our rhel6 system with openblas 0.3.3:

bash-4.1$ git status
On branch cp2k-fixes
Your branch is up-to-date with 'origin/cp2k-fixes'.

bash-4.1$ ./bin/spack install openblas
==> Installing openblas               
==> Using cached archive: /nfs/tmp2/lee218/delete/spack.pr9308/var/spack/cache/openblas/openblas-0.3.3.tar.gz                                                   
==> Staging archive: /nfs/tmp2/lee218/delete/spack.pr9308/var/spack/stage/openblas-0.3.3-7rnqrgod2pk3tlgcq6ggx7yzwiugogwv/v0.3.3.tar.gz                         
==> Created stage in /nfs/tmp2/lee218/delete/spack.pr9308/var/spack/stage/openblas-0.3.3-7rnqrgod2pk3tlgcq6ggx7yzwiugogwv                                       
==> No patches needed for openblas                                              
==> Building openblas [MakefilePackage]                                         
==> Executing phase: 'edit'                                                     
==> Executing phase: 'build'                                                    
==> Error: ProcessError: Command exited with status 2:                          
    'make' '-j16' 'CC=/nfs/tmp2/lee218/delete/spack.pr9308/lib/spack/env/gcc/gcc' 'FC=/nfs/tmp2/lee218/delete/spack.pr9308/lib/spack/env/gcc/gfortran' 'MAKE_NB_JOBS=0' 'NO_AVX2=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'libs' 'netlib' 'shared'      

77 errors found in build log:
     9057    perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 > linktest
             .c                                                                
     9058    /nfs/tmp2/lee218/delete/spack.pr9308/lib/spack/env/gcc/gcc -O2 -DM
             AX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_AVX2
              -DNO_AVX512 -DNO_WARMUP -DMAX_CPU_NUMBER=16 -DMAX_PARALLEL_NUMBER
             =1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME= -DASMFNAME=_ 
             -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINIT
             Y -I..  -shared -o ../libopenblas_sandybridge-r0.3.3.so \         
     9059       -Wl,--whole-archive ../libopenblas_sandybridge-r0.3.3.a -Wl,--no-                                                                               
             whole-archive \                                                    
     9060       -Wl,-soname,libopenblas.so.0 -lm -lgfortran -lm -lgfortran      
     9061    /nfs/tmp2/lee218/delete/spack.pr9308/lib/spack/env/gcc/gcc -O2 -DM 
             AX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_AVX2 
              -DNO_AVX512 -DNO_WARMUP -DMAX_CPU_NUMBER=16 -DMAX_PARALLEL_NUMBER 
             =1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME= -DASMFNAME=_  
             -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINIT 
             Y -I..  -w -o linktest linktest.c ../libopenblas_sandybridge-r0.3. 
             3.so -L/usr/lib/gcc/x86_64-redhat-linux6E/4.6.1 -L/usr/lib/gcc/x86 
             _64-redhat-linux6E/4.6.1/../../../../lib64 -L/lib/../lib64 -L/usr/ 
             lib/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux6E/4.6.1/../../..   
             -lgfortran -lm -lquadmath -lm -lc   && echo OK.                    
     9062    /tmp/ccvwkakE.o: In function `main':                               
  >> 9063    linktest.c:(.text.startup+0x2f4e): undefined reference to `clagge_ 
             '                                                                  
  >> 9064    linktest.c:(.text.startup+0x2f55): undefined reference to `claghe_ 

...

  >> 9138    make[1]: *** [../libopenblas_sandybridge-r0.3.3.so] Error 1        
     9139    make[1]: Leaving directory `/tmp/lee218/spack-stage/spack-stage-10 
             4EwG/OpenBLAS-0.3.3/exports'                                       
  >> 9140    make: *** [shared] Error 2                                         

See build log for details:
  /nfs/tmp2/lee218/delete/spack.pr9308/var/spack/stage/openblas-0.3.3-7rnqrgod2pk3tlgcq6ggx7yzwiugogwv/OpenBLAS-0.3.3/spack-build.out

Reply:
@lee218llnl ok, thanks. Can you upload the complete build log somewhere? For some reason some of the object files don't get linked into the `libopenblas_sandybridge-r0.3.3.so` which causes the undefined reference, but I don't see the respective line in the above snippet.